File: Exercism/haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)


properDivisors :: Int -> [Int]
properDivisors x = [y | y <- [1..x-2], x `mod` y == 0]

aliquotSum :: Int -> Int
aliquotSum = sum . properDivisors

classify :: Int -> Maybe Classification
classify x
  | x <=0 = Nothing
  | otherwise = Just $ _classify (aliquotSum x)
  where _classify asum
          | asum < x  = Deficient
          | asum == x = Perfect
          | otherwise = Abundant

File: Exercism/python/perfect-numbers/perfect_numbers.py
def cmp(x, y):
    return (x > y) - (y > x)


def classify(number):
    if number <= 0:
        raise ValueError("Cannot classify negative or zero number")
    return ["perfect", "abundant", "deficient"][
        cmp(
            sum(x for x in range(1, number - 1) if number % x == 0),
            number
        )
    ]
